Minor optimization of STObject::add

scottschurr opened the pull request:
Replace a local `std::map` in `STObject::add()` with a sorted `std::vector`.  It is possible to call `reserve()` with a sensible value on the `vector`, so this should give us significantly less freestore activity in the method.  The element types in the `vector` are cheaply copyable, so the `sort` should be inexpensive.

Also replaces two calls to `dynamic_cast` with a single virtual method call.  In any implementation I'm aware of a `dynamic_cast` is more expensive than a single virtual call.

The two added `assert`s are intended to provide confidence that the virtual call really _does_ provide identical functionality to the `dynamic_casts`.  All unit tests pass on a debug build with the asserts in place.  I've also run a debug version of the server against the network for four hours without either assert firing.  I'll probably give it 24 hours of burn in.  Assuming it survives the 24 hour burn in, I'm not convinced that the `assert`s need to be kept long term, but I'm open to opinions on that.

Serialization of `STObject`s is common and important.  It should be as fast as reasonable while still being correct.

Reviewers: @HowardHinnant @seelabs

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/4f286bdd024d819638beb6f427abf9838b16b4cb)

Built at __20180710 - 01:30:21__

### Test Results

Build Type | Log | Result | Status
---------- | --- | ------ | ------
rpm | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/rpm.txt) | 1034 cases, 0 failed, t: n/a | PASS :white_check_mark: 
msvc.debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/msvc.debug.txt) | 1032 cases, 0 failed, t: 565s | PASS :white_check_mark: 
gcc.release.unity <br/>-Dassert=true, <br/>MANUAL_TESTS=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/gcc.release.unity_Dassert_true_MANUAL_TESTS_true.txt) | 1101 cases, 0 failed, t: 796s | PASS :white_check_mark: 
docs | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/docs.txt) | 1 cases, 0 failed, t: 2s | PASS :white_check_mark: 
msvc.debug, <br/>PROJECT_NAME=rippled_classic | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/msvc.debug_PROJECT_NAME_rippled_classic.txt) | 1032 cases, 0 failed, t: 1237s | PASS :white_check_mark: 
coverage | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/coverage.txt) | 1035 cases, 0 failed, t: 690s | PASS :white_check_mark: 
clang.debug.unity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/clang.debug.unity.txt) | 1035 cases, 0 failed, t: 252s | PASS :white_check_mark: 
clang.debug.nounity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/clang.debug.nounity.txt) | 1033 cases, 0 failed, t: 148s | PASS :white_check_mark: 
msvc.release | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/msvc.release.txt) | 1031 cases, 0 failed, t: 517s | PASS :white_check_mark: 
clang.debug.unity, <br/>PARALLEL_TESTS=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/clang.debug.unity_PARALLEL_TESTS_false.txt) | 1035 cases, 0 failed, t: 490s | PASS :white_check_mark: 
gcc.debug.unity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/gcc.debug.unity.txt) | 1035 cases, 0 failed, t: 248s | PASS :white_check_mark: 
gcc.debug.nounity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/gcc.debug.nounity.txt) | 1033 cases, 0 failed, t: 110s | PASS :white_check_mark: 
clang.release.unity <br/>-Dassert=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/clang.release.unity_Dassert_true.txt) | 1034 cases, 0 failed, t: 337s | PASS :white_check_mark: 
gcc.release.unity <br/>-Dassert=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/gcc.release.unity_Dassert_true.txt) | 1034 cases, 0 failed, t: 354s | PASS :white_check_mark: 
gcc.debug.unity <br/>-Dstatic=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2621/1//artifact/gcc.debug.unity_Dstatic_true.txt) | 1035 cases, 0 failed, t: 250s | PASS :white_check_mark:

seelabs commented on src/ripple/protocol/impl/STObject.cpp:
```diff
@@ -720,10 +725,16 @@ void STObject::add (Serializer& s, bool withSigningFields) const
+        assert ((sType == STI_ARRAY) ==
+            (dynamic_cast<const STArray*> (field) != nullptr));
+        assert ((sType == STI_OBJECT) ==
+            (dynamic_cast<const STObject*> (field) != nullptr));
```
BTW, I vote to remove the asserts, but I'm fine if you want to keep them.

HowardHinnant replied:
No opinion on these asserts.  Fine either way.

seelabs commented on src/ripple/protocol/impl/STObject.cpp:
```diff
@@ -696,22 +696,27 @@ bool STObject::operator== (const STObject& obj) const
         }
     }
 
+    std::sort (entries.begin(), entries.end());
```
If you want to be really paranoid, you can check that there are no duplicate fieldCodes in the entries collection.

HowardHinnant replied:
I share both of these concerns too.  The lambda sort will also add a small bit of savings in code size (dropping a not-used compare).

Is a duplicate key a logic error or something that should be ignored?  If logic error, assert, otherwise `std::unique` could be used on the sorted vector.

HowardHinnant replied:
If assert, I believe this is the proper test:  `std::adjacent_find(entries.begin(), entries.end(), [](auto const& x, auto const& y) {return x.first == y.first;}) == entries.end()` (untested)

scottschurr commented:
Thanks very much for the comments.  I hope I've addressed them, but I took a somewhat circuitous path to do so.  There was a pre-existing private sort function in `STObject` that _almost_ did the right stuff.  So I added a parameter to that function and did a little bit of refactoring to make that all work.

BTW, I finished my 24 hours of running against the network with the two new asserts in place.  Nothing fired and I've now removed the asserts.

Please let me know what you think.  Thanks again for the careful review.

JoeLoser commented on src/ripple/protocol/STObject.h:
```diff
@@ -538,13 +538,22 @@ class STObject
 
 private:
-    void add (Serializer & s, bool withSigningFields) const;
+    enum WhichFields : bool
```
Glad you're using more-of named-parameters than just unnamed bools. This related back to your comment in a previous PR. 👍

JoeLoser commented on src/ripple/protocol/impl/STObject.cpp:
```diff
@@ -694,67 +694,61 @@ bool STObject::operator== (const STObject& obj) const
-    std::sort (sf.begin (), sf.end (),
-        [] (STBase const* a, STBase const* b) -> bool
+    std::sort (sf.begin(), sf.end(),
+        [] (STBase const* lhs, STBase const* rhs) -> bool
```
Do we need the trailing return type for the compare lambda? I think we can get away with it.

scottschurr commented on src/ripple/protocol/STObject.h:
```diff
@@ -538,13 +538,22 @@ class STObject
+        // These values are carefully chosen to do the right thing if passed
+        // to SField::shouldInclude (bool)
+        withoutSigningFields = false,
+        withAllFields = true
```
I think the distinction between `with` and `omit` is great.  I'm leaning toward
```
omitSigningFields = false,
withAllFields = true
```
I prefer `withAllFields` over `withSigningFields`, because it says what were doing.  We leave nothing out.  But that's just my take.  Will that work for you @seelabs?

seelabs replied:
I'm fine with that.

scottschurr commented on src/ripple/protocol/STObject.h:
```diff
@@ -538,13 +538,22 @@ class STObject
     static std::vector<STBase const*>
-    getSortedFields (STObject const& objToSort);
+    getSortedFields (
+        STObject const& objToSort, WhichFields whichFields);
```
I think the only reason that method is `static` is because it does not rely on the `this` parameter.  So it's leaking an implementation detail.  I think that's not too bad since it's a private method.

seelabs replied:
The `this` paramter should be `objToSort`.

scottschurr replied:
That signature could work, but it leaves me a little queasy about lifetime considerations.  For some reason the current signature gives me a better suggestion that the returned `vector` should not outlive the passed in `STObject&`.  But I'm not sure why I feel that way.

nbougalis commented:
Merged with #2629 as 574ea2c14da6349f51054e8f0a268642ffe7faea
